runtime/src/secondary_index.rs

```diff
@@ -99,12 +100,138 @@ impl SecondaryIndexEntry for RwLockSecondaryIndexEntry {
     }
 }
 
+#[derive(Debug, Default)]
+pub struct HierarchicalOrderedMap<L1, L2>
```

Comment:
I would argue that the ordered-map is not faster for this implementation. Thats why I didn't use it. It may be faster to do a re-sort, but this is specifically designed to avoid ever doing a full re-sort, or doing lots of allocs during operation. Instead it uses a binary search to always insert in the correct ordered position. Sorting requires a N*log(N) operation while a search on ordered data is only log(N). Since it's built to stay ordered at all times, I avoid the need to ever do a full resort, or needing to allocate more than once up front. Plus the ordered-map is most certainly a higher memory footprint.